### PR TITLE
Use topological sort to verify processors never run in parallel

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/Engineering/issues/1190. Depends on https://github.com/NuGet/NuGet.Jobs/pull/353.

We are adding the notion of processors to the orchestrator. A processor is simply a validator that can mutate package content during validation. This is necessary when stripping and adding repository signatures.

Processors cannot run in parallel with other processors or even with other validators. Suppose you have the following validator graph (where `-->` indicates a validator dependent)

```
                  ---> Validator B ---
                /                      \
Validator A ---                          ---> Validator D
                \                      /
                  ---> Validator C ---
```

In this case, B and C cannot be processors. A and D can. Given a graph of validator dependencies and the list of validators that are processors, I use the following algorithm to determine whether a processor can run in parallel with a validator.

1. Enumerate all valid orderings using topological sort. In our example above, this would be:
    - `A B C D`
    - `A C B D`
1. For each processor, verify that the position in all orderings is the same.
    - Note that A is always first and D is always fourth.

Alternatively, I could have enhanced the logic in orchestrator to determine whether a specific validator (or processor) could be start, giving the current set of completed and in progress validators. I decided against this since it would complicate orchestrator guts since there are now _two_ inputs for this resolution:

1. What are the dependencies of each validator? (existing)
1. Is the validator a processor? (new)

If we ever wanted to allow certain validations to run in parallel with processors or if we wanted to change something else in the validator ordering, it would mean more changes to critical parts of orchestrator. This PR basically means that orchestrator will never even start up if it is possible to have a processor in parallel with an orchestrator.

I preferred this approach of validating the configuration in advance since it allows us to determine the safety of ordering at build time (since we know the configuration at this time) rather than in the middle or orchestrator operation. As an added benefit, this also means that ordering of validations is more predictable because it is encoded into to configuration and therefore a lot easier to look up and understand.

Note that topological sort also gives us cycle detection, so I gutted that part of the existing validation.

### CODE REVIEW

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                    foreach (var dependencyName in node.DependentValidations)
+                    {
+                        graph[dependencyName].InDegree++;
+                    }
+
+                    done = true;
+                }
+            }
+
+            if (!done && currentResult.Count == graph.Count)
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Is a configuration with currentResult.Count != graph.Count valid? If not should we throw?

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: I tried to keep the topologicial sort part without exceptions. If this is the case, an empty list of results will be returned and we will throw at a higher level.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        public static void Validate(IReadOnlyList<ValidationConfigurationItem> validators, IReadOnlyList<string> cannotBeParallel)
+        {
+            var allOrders = EnumerateAll(validators);
+            if (!allOrders.Any())
+            {
+                throw new ConfigurationErrorsException("No validation sequences were found. This indicates a cycle in the validation dependencies.");
+            }
+
+            var nameToExpectedIndex = allOrders[0]
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Maybe add comments for the nameToExpectedIndex

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Sure, sounds fine.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+using System;
+using System.Collections.Generic;
+using System.Configuration;
+using System.Linq;
+
+namespace NuGet.Services.Validation.Orchestrator
+{
+    public static class TopologicalSort
+    {
+        public static void Validate(IReadOnlyList<ValidationConfigurationItem> validators, IReadOnlyList<string> cannotBeParallel)
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: I think that a cyclic graph like 1->2->3->1 will pass the validation.

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Why do you think that? All of these nodes will have an in-degree of 1 and therefore the topological sort method will never recurse, resulting in an empty result set.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Fixed.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Added this case to the test suite. Works as expected (cycle is detected).

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                .Select((x, i) => new { Name = x, Index = i })
+                .ToDictionary(x => x.Name, x => x.Index);
+
+            foreach (var order in allOrders.Skip(1))
+            {
+                foreach (var name in cannotBeParallel)
+                {
+                    var index = nameToExpectedIndex[name];
+                    var otherName = order[nameToExpectedIndex[name]];
+                    if (otherName != name)
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: maybe use the index at line 29? ( var otherName = order[index];)

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Good call.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            var nameToExpectedIndex = allOrders[0]
+                .Select((x, i) => new { Name = x, Index = i })
+                .ToDictionary(x => x.Name, x => x.Index);
+
+            foreach (var order in allOrders.Skip(1))
+            {
+                foreach (var name in cannotBeParallel)
+                {
+                    var index = nameToExpectedIndex[name];
+                    var otherName = order[nameToExpectedIndex[name]];
```
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: cristinamanum just now  Owner
maybe use the index at line 29? ( var otherName = order[index];)

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Fixed.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs (hunk 3 above)
Comment (cristinamanum) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Maybe add comments on the Validate method? (maybe part of the PR comments?)

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Sure

Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: Fixed.

Comment (cristinamanum) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/TopologicalSort.cs]: You are correct, I missed the validation at line 16.

**Pull request merged**